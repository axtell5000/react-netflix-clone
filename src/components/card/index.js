import React, { useState, useContext, createContext } from 'react';

import {
  Container,
  Group,
  Title,
  SubTitle,
  Text,
  Feature,
  FEatureTitle,
  FeatureText,
  FeatureClose,
  Maturity,
  Content,
  Entities,
  Item,
  Image
} from './styles/card';

export const FeatureContext = createContext();

export default function Card({ children, ...restProps }) {
  // some state
  const [showFeature, setShowFeature] = useState(false);
  const [itemFeature, setItemFeature] = useState({});

  return (
    // setting up so this part of context goes to this part of App
    <FeatureContext.Provider value={{ showFeature, setShowFeature, itemFeature, setItemFeature }}>
      <Container {...restProps}>{children}</Container>
    </FeatureContext.Provider>
  );
}

// Holds all the cards
Card.Group = function CardGroup({ children, ...restProps }) {
  return <Group {...restProps}>{children}</Group>;
};

Card.Title = function CardTitle({ children, ...restProps }) {
  return <Title {...restProps}>{children}</Title>;
};

Card.SubTitle = function CardSubTitle({ children, ...restProps }) {
  return <SubTitle {...restProps}>{children}</SubTitle>;
};

Card.Text = function CardText({ children, ...restProps }) {
  return <Text {...restProps}>{children}</Text>;
};

// Hold information about series / film
Card.Meta = function CardMeta({ children, ...restProps }) {
  return <Meta {...restProps}>{children}</Meta>;
};

Card.Entities = function CardEntities({ children, ...restProps }) {
  return <Entities {...restProps}>{children}</Entities>;
};

Card.Item = function CardItem({ item ,children, ...restProps }) {
  // pulling state from the context
  const { setShowFeature, setItemFeature } = useContext(FeatureContext);
  
  return (
    <Item
      onClick={() => {
        setItemFeature(item);
        setShowFeature(true);
      }}
      {...restProps}
    >
      {children}
    </Item>
  );
};

Card.Image = function CardImage({ ...restProps }) {
  return <Image {...restProps} />;
};