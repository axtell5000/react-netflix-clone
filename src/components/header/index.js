import React, { useState } from 'react';
import { Link as ReachRouterLink } from 'react-router-dom';

import { 
  Background, 
  Container, 
  Dropdown, 
  Profile, 
  ButtonLink, 
  Logo, 
  Feature, 
  Text, 
  FeatureCallOut, 
  Link, 
  Group, 
  Picture,
  Search,
  SearchIcon,
  SearchInput,
  PlayButton
} from './styles/header';

export default function Header({ background = true, children, ...restProps}) {
  return (
    background ? <Background {...restProps}>{children}</Background>: children
  );
}

Header.Frame = function HeaderFrame({ children, ...restProps }) {
  return <Container {...restProps}>{children}</Container>;
};

Header.Group = function HeaderGroup({ children, ...restProps }) {
  return <Group {...restProps}>{children}</Group>;
};


Header.ButtonLink = function HeaderButtonLink({ children, ...restProps }) {
  return <ButtonLink {...restProps}>{children}</ButtonLink>;
};

Header.Logo = function HeaderLogo({ to, ...restProps }) {
  return (
    <ReachRouterLink to={to}>
      <Logo {...restProps} />
    </ReachRouterLink>
  );
};

Header.Feature = function HeaderFeature({ children, ...restProps }) {
  return <Feature>{children}</Feature>;
};

Header.Profile = function HeaderProfile({ children, ...restProps }) {
  return <Profile {...restProps}>{children}</Profile>;
};

Header.Search = function HeaderSearch({ searchTerm, setSearchTerm, ...restProps }) {
  const [searchActive, setSearchActive] = useState(false);

  return (
    <Search {...restProps}>
      <SearchIcon onClick={() => setSearchActive((searchActive) => !searchActive)}>
        <img src="/images/icons/search.png" alt="Search" />
      </SearchIcon>
      <SearchInput
        value={searchTerm}
        onChange={({ target }) => setSearchTerm(target.value)}
        placeholder="Search films and series"
        active={searchActive}
        data-testid="search-input"
      />
    </Search>
  );
};


Header.FeatureCallOut = function HeaderFeatureCallOut({ children, ...restProps }) {
  return <FeatureCallOut>{children}</FeatureCallOut>;
};

Header.Text = function HeaderText({ children, ...restProps }) {
  return <Text {...restProps}>{children}</Text>;
};

Header.TextLink = function HeaderTextLink({ children, ...restProps }) {
  return <Link {...restProps}>{children}</Link>;
};

Header.Picture = function HeaderPicture({ src, ...restProps }) {
  return <Picture {...restProps} src={`/images/users/${src}.png`} />;
};

Header.Dropdown = function HeaderDropdown({ children, ...restProps }) {
  return <Dropdown {...restProps}>{children}</Dropdown>;
};

Header.PlayButton = function HeaderPlayButton({ children, ...restProps }) {
  return <PlayButton {...restProps}>{children}</PlayButton>;
};
