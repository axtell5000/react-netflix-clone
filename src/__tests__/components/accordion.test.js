import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import faqsData from '../../fixtures/faqs';
import { Accordion } from '../../components';

describe('<Accordion />', () => {
  it('renders the <Accordion /> with populated data', () => {
    const { container, getByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        <Accordion.Frame>
          {faqsData.map((item) => (
            <Accordion.Item key={item.id}>
              <Accordion.Header>{item.header}</Accordion.Header>
              <Accordion.Body>{item.body}</Accordion.Body>
            </Accordion.Item>
          ))}
        </Accordion.Frame>
      </Accordion>
    );
    // here checking if text is there in snapshot
    expect(getByText('Frequently Asked Questions')).toBeTruthy();
    expect(getByText('What is Netflix?')).toBeTruthy();
    expect(getByText('How much does Netflix cost?')).toBeTruthy();
    expect(getByText('Where can I watch?')).toBeTruthy();
    expect(getByText('How do I cancel?')).toBeTruthy();
    expect(getByText('What can I watch on Netflix?')).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('opens and closes the <Accordion /> component', () => {
    const { container, queryByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        <Accordion.Frame>
          {faqsData.map((item) => (
            <Accordion.Item key={item.id}>
              <Accordion.Header>{item.header}</Accordion.Header>
              <Accordion.Body data-testid="accordion-body">{item.body}</Accordion.Body>
            </Accordion.Item>
          ))}
        </Accordion.Frame>
      </Accordion>
    );

    const whatIsNetflixBodyText =
      "Netflix is a streaming service that offers a wide variety of award-winning TV programmes, films, anime, documentaries and more – on thousands of internet-connected devices. You can watch as much as you want, whenever you want, without a single advert – all for one low monthly price. There's always something new to discover, and new TV programmes and films are added every week!";

    expect(queryByText(whatIsNetflixBodyText)).toBeFalsy(); // should be false because text is not there
    fireEvent.click(queryByText('What is Netflix?')); // simulate a click on text 'What is Netflix?'
    expect(queryByText(whatIsNetflixBodyText)).toBeTruthy(); // now text should be there

    fireEvent.click(queryByText('What is Netflix?')); // fire click event to toggle offf
    expect(queryByText(whatIsNetflixBodyText)).toBeFalsy(); // now text shouldnt be there
    expect(container.firstChild).toMatchSnapshot();
  });
});
