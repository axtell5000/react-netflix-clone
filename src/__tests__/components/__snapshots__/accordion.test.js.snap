// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> opens and closes the <Accordion /> component 1`] = `
<div
  class="accordion__Container-l3iym9-0 kmHLsi"
>
  <div
    class="accordion__Inner-l3iym9-2 fdqUAe"
  >
    <h1
      class="accordion__Title-l3iym9-3 hHHSbm"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="accordion__Frame-l3iym9-1 lhGwav"
    >
      <div
        class="accordion__Item-l3iym9-4 dDNVMV"
      >
        <div
          class="accordion__Header-l3iym9-5 gxafAw"
        >
          What is Netflix?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
      </div>
      <div
        class="accordion__Item-l3iym9-4 dDNVMV"
      >
        <div
          class="accordion__Header-l3iym9-5 gxafAw"
        >
          How much does Netflix cost?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
      </div>
      <div
        class="accordion__Item-l3iym9-4 dDNVMV"
      >
        <div
          class="accordion__Header-l3iym9-5 gxafAw"
        >
          Where can I watch?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
      </div>
      <div
        class="accordion__Item-l3iym9-4 dDNVMV"
      >
        <div
          class="accordion__Header-l3iym9-5 gxafAw"
        >
          How do I cancel?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
      </div>
      <div
        class="accordion__Item-l3iym9-4 dDNVMV"
      >
        <div
          class="accordion__Header-l3iym9-5 gxafAw"
        >
          What can I watch on Netflix?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Accordion /> renders the <Accordion /> with populated data 1`] = `
<div
  class="accordion__Container-l3iym9-0 kmHLsi"
>
  <div
    class="accordion__Inner-l3iym9-2 fdqUAe"
  >
    <h1
      class="accordion__Title-l3iym9-3 hHHSbm"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="accordion__Frame-l3iym9-1 lhGwav"
    >
      <div
        class="accordion__Item-l3iym9-4 dDNVMV"
      >
        <div
          class="accordion__Header-l3iym9-5 gxafAw"
        >
          What is Netflix?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
      </div>
      <div
        class="accordion__Item-l3iym9-4 dDNVMV"
      >
        <div
          class="accordion__Header-l3iym9-5 gxafAw"
        >
          How much does Netflix cost?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
      </div>
      <div
        class="accordion__Item-l3iym9-4 dDNVMV"
      >
        <div
          class="accordion__Header-l3iym9-5 gxafAw"
        >
          Where can I watch?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
      </div>
      <div
        class="accordion__Item-l3iym9-4 dDNVMV"
      >
        <div
          class="accordion__Header-l3iym9-5 gxafAw"
        >
          How do I cancel?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
      </div>
      <div
        class="accordion__Item-l3iym9-4 dDNVMV"
      >
        <div
          class="accordion__Header-l3iym9-5 gxafAw"
        >
          What can I watch on Netflix?
          <img
            alt="Open"
            src="/images/icons/add.png"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
